The ANTLR-isation of the lexical grammar. The C# "lexical" grammar bo…

Nigel-Ecma opened the pull request:
…th tokenises and parses the input and this

requires using both lexical and parser ANTLR rules while producing a lexer which produces a token stream and
not some AST of the input!

In this stage of ANTLR-isation the pre-processor is not tackled; the pre-processor definition is a context-free
grammar which is atypical for a lexical grammar and defining it with lexical rules does not produce the expected
result... That said the semantics of the pre-processor can be acheived in an ANTLR lexical grammar which includes
predicates/actions and that will be explored in a subsequent stage. The grammar in this stage can be used along
with a factored syntactical grammar to lex & parse C# source which does not include pre-processor constructs.

The majority of changes are to lexical-structure.md, many of the other files are only altered as the lexical
rule *Identifier* has been replaced by the parser rule *identifier*.

I will add some comments within the PR to explain why it is the way it is...

Nigel-Ecma committed:
The ANTLR-isation of the lexical grammar. The C# "lexical" grammar both tokenises and parses the input and this
requires using both lexical and parser ANTLR rules while producing a lexer which produces a token stream and
not some AST of the input!

In this stage of ANTLR-isation the pre-processor is not tackled; the pre-processor definition is a context-free
grammar which is atypical for a lexical grammar and defining it with lexical rules does not produce the expected
result... That said the semantics of the pre-processor can be acheived in an ANTLR lexical grammar which includes
predicates/actions and that will be explored in a subsequent stage. The grammar in this stage can be used along
with a factored syntactical grammar to lex & parse C# source which does not include pre-processor constructs.

The majority of changes are to lexical-structure.md, many of the other files are only altered as the lexical
rule *Identifier* has been replaced by the parser rule *identifier*.

Nigel-Ecma commented on standard/grammar.md:
```diff
@@ -19,23 +19,23 @@ ASTERISK : '*' ;
 // Source: §7.3.1 General
-Input
-    : Input_Section?
+input
```
This file is largely auto-generated on GitHub, the changes here (a) reflect changes elsewhere and (b) are incomplete. Once the GitHub tooling runs they will be replaced. In other words don't review this file!

Nigel-Ecma commented on standard/lexical-structure.md:
```diff
@@ -26,23 +26,29 @@ All terminal characters are to be understood as the appropriate Unicode characte
-The lexical and syntactic grammars are presented in Backus-Naur form using the notation of the ANTLR grammar tool.
+The lexical and syntactic grammars are presented in the ANTLR grammar tool’s Extended Backus-Naur form.
+
+While the ANTLR notation is used this Standard does not present a complete ANTLR-ready "reference grammar" for C#; writing a lexer and parser, either by hand or using a tool such as ANTLR, is outside the scope of a language specification. With that qualification, this Standard attempts to minmise the gap between the specified grammar and that required to build a lexer and parser in ANTLR with the ...
```
Normative, informative, or a mix?

jskeet replied:
Nit: minmise => minimize

I think it's reasonable to keep this part normative. Will review the mix as I go along.

Nigel-Ecma replied:
Catching my typos is not a nit ;-) Fixed.

Nigel-Ecma commented on standard/lexical-structure.md:
```diff
@@ -26,23 +26,29 @@ All terminal characters are to be understood as the appropriate Unicode characte
+
+ANTLR distinguishes between lexical and syntactic, termed parser by ANTLR, grammars in its notation by starting lexical rules with an initial uppercase letter and parser rules with an initial lowercase letter.
+
+*Note*: The C# lexical grammar ([§7.2.3](lexical-structure.md#723-lexical-grammar)) and syntactic grammar ([§7.2.4](lexical-structure.md#724-syntactic-grammar)) are not in exact correspondence with the ANTLR division into lexical and parser grammers. This small mismatch means that some ANTLR parser rules are used when specifying the C# lexical grammar. *end note* 
```
Or put another way the C# "lexical" grammar is a context-free grammar not a regular one... Nothing wrong with that per se but it's not the model a tool like ANTLR follows.

I've added a number of notes (which are informative rather than normative) thoughout, the number & content of these should be reviewed.

Nigel-Ecma commented on standard/lexical-structure.md:
```diff
@@ -26,23 +26,29 @@ All terminal characters are to be understood as the appropriate Unicode characte
 The lexical grammar of C# is presented in [§7.3](lexical-structure.md#73-lexical-analysis), [§7.4](lexical-structure.md#74-tokens), and [§7.5](lexical-structure.md#75-pre-processing-directives). The terminal symbols of the lexical grammar are the characters of the Unicode character set, and the lexical grammar specifies how characters are combined to form tokens ([§7.4](lexical-structure.md#74-to ...
 
-Every compilation unit in a C# program shall conform to the *Input* production of the lexical grammar ([§7.3.1](lexical-structure.md#731-general)).
+Many of the terminal symbols of the syntactic grammar are not defined explicitly as tokens in the lexical grammar. Rather advantage is taken of the ANTLR behaviour that literal strings in the grammar are extracted as implicit lexical tokens; this allows keywords, operators, etc. to be represented in the grammar by their literal representation rather than a token name.
```
That literal strings in the grammar become implicit lexical tokens was also the case with the original grammar. The impact of this was never spelled out in any way and indeed the lexical rules around identifiers were described as though the implicitly defined keyword tokens didn't exist. As we are now using the language of a real parser generator this detail has been added.

jskeet replied:
Nit: behavior => behavior (it pains me, but...) - same later on.

Nigel-Ecma replied:
Yup, I'm used to correct (British/New Zealand) spelling ;-)

Nigel-Ecma commented on standard/lexical-structure.md:
````diff
@@ -99,31 +105,33 @@ ASTERISK : '*' ;
     ;
 ```
 
+> *Note*: The above grammar is described by ANTLR parsing rules, it defines the lexical structure of a C# compilation unit and not lexical tokens. *end note*
````
Really this grammar is not about tokens (lexical) but about arrangements of tokens (parser). A lexer might implement them to check the structure of the input but there is no need and in ANTLR they are simply ignored as there is no path from the start symbol of the parser grammar to them. However they are the only place that *PP_Directive* is referenced, but that is no ANTLR-ised yet...

jskeet replied:
Both the standard change and the comment have gone so far over my head, they've left earth's atmosphere. I think that's fine - but if you expect a useful review of this little bit, please let me know :)

Nigel-Ecma commented on standard/lexical-structure.md:
```diff
@@ -132,6 +140,12 @@ When several lexical grammar productions match a sequence of characters in a com
 
 > *Example*: The character sequence `//` is processed as the beginning of a single-line comment because that lexical element is longer than a single `/` token. *end example*
 
+The rules of the lexical grammar are ordered top-to-bottom, with the first matching rule determining the recognised token. Implicit rules ([§7.2.3](lexical-structure.md#723-lexical-grammar)) derived from literal strings in the grammar are ordered before the explicit rules. This ordering simplifies the grammar.
```
This top-to-bottom ordering is not mentioned in the original grammar but is how lexical grammars usually work – lexical rules are ordered while parser rules form a tree rooted in a start symbol. It is important to mention it once the implicit rules for keywords etc. are introduced.

jskeet replied:
+1 for mentioning it explicitly.

Nigel-Ecma commented on standard/lexical-structure.md:
````diff
@@ -324,64 +340,94 @@ Multiple translations are not performed. For instance, the string literal `"\u00
 
 ```ANTLR
-Identifier
+identifier
````
This part of the grammar has been re-formulated noting (a) the impact of the implicit rules generated from literal strings in the grammar and (b) to leverage those so "contextual keywords" are recognised *by the grammar* as identifiers when not being contextually – there might be cases when a contextual keyword is parsed as an identifier even though its in a contextual position but that can easily be handled by the compiler.

Nigel-Ecma commented on standard/lexical-structure.md:
```diff
@@ -324,64 +340,94 @@ Multiple translations are not performed. For instance, the string literal `"\u00
+    | contextual_keyword
+    ;
+
+Simple_Identifier
```
You might wonder why *Simple_Identifier* exists and its simply not folded into *identifier*. So that the main lexical rules can be represented as a lexical grammar, rather than a combined grammar, in ANTLR the C# "lexical" rules are split into two ANTLR grammars (a) a lexical one for lexical rules and (b) a parser one for the parser rules. This arrangement will allow the pre-processor to be implemented as such an implementation is likely to rely on features only ANTLR lexical-only grammars can use. The rule `Simple_Identifer` provides the bridge between the lexer and parser rules of the "lexical" grammar... Now you know why I didn't explain that in the text, too tortuous! And its not as though the grammar doesn't have plenty of redundant rules for the purposes of exposition already, and the next one is immediately below...

Nigel-Ecma commented on standard/lexical-structure.md:
```diff
@@ -324,64 +340,94 @@ Multiple translations are not performed. For instance, the string literal `"\u00
-    | Unicode_Escape_Sequence { IsLetterCharacter() }?
+fragment Letter_Character
+    : [\p{L}\p{Nl}]           // category Letter, all subcategories; category Number, subcategory letter
+    | Unicode_Escape_Sequence // only escapes for categories L & Nl allowed, see note below
```
Should we include the semantic predicates? I'm in two (at least) minds, so we get both a comment (as in the original), a predicate, and a note below explaining they are not normative...

Nigel-Ecma commented on standard/lexical-structure.md:
```diff
@@ -324,64 +340,94 @@ Multiple translations are not performed. For instance, the string literal `"\u00
+> 2. The ANTLR semantic predicates above: `IsLetterCharacter`, `IsCombiningCharacter`, `IsDecimalDigitCharacter`, `IsConnectingCharacter` and `IsFormattingCharacter`; are informative *only*. How a compiler enforces the restriction on the allowable
+*Unicode_Escape_Sequence* values is an implementation issue. *end note*
+
+> 3. The explicit rules for identifiers depend on the implicit rules introduced by literal strings in the grammar:
```
This attempts to clarify how the implicit and explicit rules interact to tokenise identifiers and [contextual] keywords, does it? Please review carefully.

Nigel-Ecma commented on standard/lexical-structure.md:
````diff
@@ -758,26 +813,28 @@ There are several kinds of operators and punctuators. Operators are used in expr
 ```
 
-*Right_Shift* is made up of the two tokens `>` and `>`. Similarly, *Right_Shift_Assignment* is made up of the two tokens `>` and `>=`. Unlike other productions in the syntactic grammar, no characters of any kind (not even whitespace) are allowed between the two tokens in each of these productions. These productions are treated specially in order to enable the correct handling of *type_parameter_l ...
+> *Note*: `right_shift` and `right_shift_assignment` are parser rules as they do not introduce a new token kind but represent a sequence of two tokens. The `operator_or_punctuator` rule exists for descriptive purposes only and is not used elsewhere in the grammar. *end note*
````
Not said is that in the original grammar these two rules didn't really work as they would both have produced a single token...

Nigel-Ecma commented on standard/lexical-structure.md:
```diff
@@ -787,14 +844,16 @@ The pre-processing directives provide the ability to skip conditionally sections
 
 > *Note*: The term "pre-processing directives" is used only for consistency with the C and C++ programming languages. In C#, there is no separate pre-processing step; pre-processing directives are processed as part of the lexical analysis phase. *end note*
 
+> *Note*: The pre-processor grammar, though defined using ANTLR lexical rules, does not define any lexical tokens in the language and is not “ANTLR ready”. Pre-processing involves processing the input and modifying the token stream produced by the lexer, this is not represented in this grammar and is part of compiler implementation and thus out of scope for this language specification. *end note*
```
This is the big bit that hasn't been ANTLR-ised. I did look as to whether these rules could be treated as lexical ones which produce a single "PP Directive" token, but that is thwarted by the context-free specification...

Though I think I have a proposal as to how to handle this as any PR will build on this one and I've no idea how to do that in GitHub I've chosen to leave this until we've decided whether to accept this PR and its merged into the main draft-v6 branch. If you would like to know that proposal to help decide on *this* PR request it on the ECMA list and I'll explain once I've tested it.

jskeet replied:
Suggest replacing the curly-double-quotes with straight ones. But other than that, I'm fine with the paragraph and your comment :)

RexJaeschke commented:
@Nigel-Ecma I looked over this, but given my lack of background w.r.t formal grammars, I'm not following a lot of the explanation. As such, I removed myself from the reviewer's list.

BTW, I think that in at least one place you have a Note that *doesn't* start on a new line with a preceding `>`. We previously agreed that Notes can no longer be at the end of ordinary paragraphs.

Nigel-Ecma committed:
Fixed issue upsetting the automatic grammar extractor (I hope). Corrected Notes violating our style guidelines.

a) There are three rule groups; with primary rules *Input*, *Token* and *PP_Directive*; in the original "lexical"
grammar which together do not define lexical tokens but rather the lexical structure and pre-proccesing directives.
As is these rules are invalid as ANTLR lexical rules, and just switching them to be ANTLR parsing rules doesn't
work either.

In the first commit I had ANTLR-ised *Input* & *Token* groups, as the parser rule groups *input* and *token* but
had left the pre-processor rules as future work. However Bill's grammar extractor pointed out I'd violated an
ANTLR rule in doing so. I've made a small change to try to beat the grammar extractor, we'll see.

As before the pre-processor grammar remains invalid as ANTLR (and it was really always invalid as lexer rules)
but as a *description* it does its job as it always has.

(If people want the technical details ask on the ECMA list and I'll explain.)

b) Rex spotted I've violated our style guidelines for *Notes* by having some as "in paragraph". These have been
changed to be standalone indent paras.

Nigel-Ecma commented:
@RexJaeschke – I think I’ve fixed all the inline Notes that I added (there was more than one); in looking for them I did a global search and turned up two elsewhere. I have not fixed them in the draft PR as they are not connected:

basic-concepts.md, line 712
conversions.md, line 444

> On 21/06/2021, at 8:05 am, Rex Jaeschke ***@***.***> wrote:
> 
> 
> @Nigel-Ecma <https://github.com/Nigel-Ecma> I looked over this, but given my lack of background w.r.t formal grammars, I'm not following a lot of the explanation. As such, I removed myself from the reviewer's list.
> 
> BTW, I think that in at least one place you have a Note that doesn't start on a new line with a preceding >. We previously agreed that Notes can no longer be at the end of ordinary paragraphs.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/dotnet/csharpstandard/pull/339#issuecomment-864605028>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABSYVW6XQB2MPO43SUIC5U3TTZCZZANCNFSM466UCA5A>.
>

Nigel-Ecma committed:
This commit effectively disables the parts of the lexical grammar which are not acceptable to ANTLR

a) The `Input` group of rules which define the lexical structure rather than tokens

b) The `PP_Directive` group of rules which define the pre-processor directives. The rules as is
   define also of these as single tokens, which isn't what is intended!

c) The `Token` rule, and associated introduced `Any*` rules, which are used only by (a) and (b)

The disabling is accomplished by a liberal spreading of `fragment`.

In phase 2 the intention is to convert the pre-processor description to be compatible with ANTLR.

Nigel-Ecma committed:
Oops, missed a change from the last commit, mea culpa :-(

jskeet commented on standard/lexical-structure.md:
```diff
@@ -26,23 +26,29 @@ All terminal characters are to be understood as the appropriate Unicode characte
 ### 7.2.2 Grammar notation
 
-The lexical and syntactic grammars are presented in Backus-Naur form using the notation of the ANTLR grammar tool.
+The lexical and syntactic grammars are presented in the ANTLR grammar tool’s Extended Backus-Naur form.
```
Are we using curly quotes in the standard in general? I'd normally expect an ASCII `'`.

Nigel-Ecma replied:
All the quotes (some historical) in *this* file have now been straightened, I leave the rest of the Standard to others. I trust that the Word convertor will put them back in.

jskeet commented on standard/lexical-structure.md:
```diff
@@ -324,64 +340,94 @@ Multiple translations are not performed. For instance, the string literal `"\u00
+
+> 1. For information on the Unicode character classes mentioned above, see *The Unicode Standard*. *end note*
+
+> 2. The ANTLR semantic predicates above: `IsLetterCharacter`, `IsCombiningCharacter`, `IsDecimalDigitCharacter`, `IsConnectingCharacter` and `IsFormattingCharacter`; are informative *only*. How a compiler enforces the restriction on the allowable
```
I'd put that list of predicates into parentheses instead:

```text
The ANTLR semantic predicates above (`IsLetterCharacter`, `IsCombiningCharacter`, `IsDecimalDigitCharacter`, `IsConnectingCharacter` and `IsFormattingCharacter`) are informative *only*.
```

Nigel-Ecma replied:
Done

jskeet commented on standard/lexical-structure.md:
```diff
@@ -324,64 +340,94 @@ Multiple translations are not performed. For instance, the string literal `"\u00
+*Unicode_Escape_Sequence* values is an implementation issue. *end note*
+
+> 3. The explicit rules for identifiers depend on the implicit rules introduced by literal strings in the grammar:
+>> - Keywords and contextual keywords occur in the grammar as literal strings and from these implicit lexical token rules are created which are ordered before the explicit lexical rules ([§7.2.3](lexical-structure.md#723-lexical-grammar)).
```
What's the purpose of the `>>` here? That's Markdown I haven't seen before.

jskeet replied:
In terms of the text itself, I think I'd find this easier to read as three sentences:

> Keywords and contextual keywords occur in the grammar as literal strings. Implicit lexical token rules are created from these literal strings. The implicit rules are ordered before the explicit lexical rules ([§7.2.3](lexical-structure.md#723-lexical-grammar)).

Nigel-Ecma replied:
>> is double-indent, saves you having to insert spaces to produce nested lists. I've switched to a single > and more spaces assuming that's the approach elsewhere. Switch to 3 sentences but started the last with "These" as it read better to me.

Nigel-Ecma replied:
And the double indent/quote worked in the comment ;-)

jskeet commented on standard/lexical-structure.md:
```diff
@@ -324,64 +340,94 @@ Multiple translations are not performed. For instance, the string literal `"\u00
+
+> 3. The explicit rules for identifiers depend on the implicit rules introduced by literal strings in the grammar:
+>> - Keywords and contextual keywords occur in the grammar as literal strings and from these implicit lexical token rules are created which are ordered before the explicit lexical rules ([§7.2.3](lexical-structure.md#723-lexical-grammar)).
+>> - Therefore fragment `Available_Identifier` will not match keywords or contextual keywords as the lexical rules for those precede it.
```
Having "Therefore" at the start of the bullet feels a little odd to me. How about

"The ordering of the token rules means that fragment [...]"

?

Nigel-Ecma replied:
We start bullets with "Otherwise", I think I'm going to disagree on this one, let's see where the consensus lands.

jskeet commented on standard/lexical-structure.md:
```diff
@@ -324,64 +340,94 @@ Multiple translations are not performed. For instance, the string literal `"\u00
+>> - Keywords and contextual keywords occur in the grammar as literal strings and from these implicit lexical token rules are created which are ordered before the explicit lexical rules ([§7.2.3](lexical-structure.md#723-lexical-grammar)).
+>> - Therefore fragment `Available_Identifier` will not match keywords or contextual keywords as the lexical rules for those precede it.
+>> - Fragment `Escaped_Identifier` will effectively include keywords and contextual keywords as part of the longer token starting with an `@`.
+>> - `identifier` is a parser rule, as are those for `keyword` and `contextual_keyword` ([§7.4.4](lexical-structure.md#744-keywords)), as they do not define a new token kinds but rather represent a grouping of other tokens.
```
Would "as are `keyword` and `contextual_keyword`" work just as well? It's not clear to me what the "those for" means or why it's different for `keyword`/`contextual_keyword` vs `identifier`.

Nigel-Ecma replied:
Dropped

jskeet commented on standard/lexical-structure.md:
````diff
@@ -459,6 +505,8 @@ Contextual_Keyword
     ;
 ```
 
+> *Note*: The rules `keyword` and `contextual_keyword` are parser rules as they do not introduce new token kinds. Every keyword and contextual keyword is defined by an implicit lexical rule as they occur as literal strings in the grammar ([§7.2.3](lexical-structure.md#723-lexical-grammar)). *end note*
````
Suggested pluralization consistency:

"All keywords and contextual keywords are defined by implicit lexical rules as they occur as literal strings in the grammar"

Nigel-Ecma replied:
Done

jskeet commented on standard/lexical-structure.md:
```diff
@@ -758,26 +813,28 @@ There are several kinds of operators and punctuators. Operators are used in expr
+*right_shift* is made up of the two tokens `>` and `>`. Similarly, *right_shift_assignment* is made up of the two tokens `>` and `>=`. Unlike other productions in the syntactic grammar, no characters of any kind (not even whitespace) are allowed between the two tokens in each of these productions. These productions are treated specially in order to enable the correct handling of *type_parameter_l ...
 
-> *Note*: Prior to the addition of generics to C#, `>>` and `>>=` were both single tokens. However, the syntax for generics uses the `<` and `>` characters to delimit type parameters and type arguments. It is often desirable to use nested constructed types, such as `List<Dictionary<string`, `int>>`. Rather than requiring the programmer to separate the `>` and `>` by a space, the definition of the ...
+> *Note*: Prior to the addition of generics to C#, `>>` and `>>=` were both single tokens. However, the syntax for generics uses the `<` and `>` characters to delimit type parameters and type arguments. It is often desirable to use nested constructed types, such as `List<Dictionary<string`, `int>>`. Rather than requiring the programmer to separate the `>` and `>` by a space, the definition of the ...
```
I don't think we want the backticks within the constructed type here - this is a bug in the existing Markdown which we didn't catch in previous review.

Nigel-Ecma replied:
Well spotted, yes an old typo, fixed.

jskeet commented on standard/lexical-structure.md:
````diff
@@ -1200,16 +1259,16 @@ The `#pragma` preprocessing directive is used to specify contextual information
 ```
 
-The *Input_Character*s in the *Pp_Pragma-Text* are interpreted by the compiler in an implementation-defined manner. The information supplied in a `#pragma` directive shall not change program semantics. A `#pragma` directive shall only change compiler behavior that is outside the scope of this language specification. If the compiler cannot interpret the *Input_Character*s, the compiler can produce ...
+The *Input_Character*s in the *PP_Pragma-Text* are interpreted by the compiler in an implementation-defined manner. The information supplied in a `#pragma` directive shall not change program semantics. A `#pragma` directive shall only change compiler behavior that is outside the scope of this language specification. If the compiler cannot interpret the *Input_Character*s, the compiler can produce ...
````
Should the - in *PP_Pragma-Text* be an underscore?

Nigel-Ecma replied:
I spotted this pre-existing typo when doing #342, I'll fix it here as well. There might be other fixes I found and made in #342, someone with good Git-fu I'm sure could merge them if we go with #339. I'm making all the fixes you found in both branches.

jskeet commented on standard/lexical-structure.md:
```diff
@@ -144,7 +146,9 @@ The rules of the lexical grammar are ordered top-to-bottom, with the first match
 > *Example*: As keyword tokens are defined implicitly by literal strings in the grammar, and implicit rules are ordered before explicit ones, the rule `Available_Identifier` ([§7.4.3](lexical-structure.md#743-identifiers)) will never match a keyword. *end example*
 
-Some tokens are defined by a set of lexical rules; a main rule and one or more sub-rules. The latter are marked in the grammer by `fragment` to indicate the rule does defines part of another token. Fragment rules are not considered in the top-to-bottom ordering of lexical rules. (*Note*: In ANTLR `fragment` is a keyword which produces the same behaviour defined here. *end note*)
+Some tokens are defined by a set of lexical rules; a main rule and one or more sub-rules. The latter are marked in the grammer by `fragment` to indicate the rule does defines part of another token. Fragment rules are not considered in the top-to-bottom ordering of lexical rules.
```
Nit: grammer => grammar

Nigel-Ecma replied:
Corrected

jskeet commented on standard/lexical-structure.md:
```diff
@@ -285,18 +291,50 @@ Whitespace
+    | 'volatile'   | 'where'     | 'while'      | 'yield'
+    ;
+
+fragment Any_Operator_Or_Punctuator
```
I seem to remember a reference to Operator_Or_Punctuator in a note, but I can't see a change to that note here.

jskeet commented on standard/lexical-structure.md:
````diff
@@ -909,10 +947,15 @@ Pre-processing directives are not tokens and are not part of the syntactic gramm
     ;
 ```
+
+> *Note* The ANTLR semantic predicate above: `IsNotTrueOrFalse`; is informative *only*. How a compiler enforces the restriction on the allowable
````
As before, I'd reword this a bit:

"The ANTLR semantic predicate above (`IsNotTrueOrFalse`) is informative *only*."

Nigel-Ecma replied:
Changed

jskeet reviewed (commented):
I'm not really qualified to comment on the validity of the overall proposal, but I have complete confidence in Nigel's efforts given his description in the previous meeting.

Not being qualified hasn't stopped me from adding various nit-picks and suggestions though :)

Nigel-Ecma committed:
Various tweaks thanks to @jskeet:
- typos
- removal of proper (British) spellings ;-)
- improved phrasing
- straightened all the quotes in lexical-structure.md *only*, checking across the whole Standard I leave to someone else. I hope the Word conversion puts them back in
- etc.

Nigel-Ecma commented:
Supplanted by #342
